Enable compiler warnings in xpack core plugin project

vamuzumd opened the pull request:
<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->
This commit is part of issue #[40366](https://github.com/elastic/elasticsearch/issues/40366) to remove disabled lint warnings from gradle files. This PR renables Xlint warnings for XPack/plugin/core project by:-

- Adding wildcard type parameters
- Suppressing unchecked warnings for most occurrences as they involved converting Object into Collection(List, Map) within lambda function.
- Removed dead url -// https://github.com/elastic/x-plugins/issues/724
- Suppressing rawtype warnings for ExecutableChainTransform class as it returns derived class object having more than 1 derived types.
- Adding unchecked suppressions for randomFrom method usages as it returns a generic object.

vamuzumd committed:
Added unchecked suppression, resolved some rawtypes

vamuzumd committed:
removed more warnings

vamuzumd commented on client/rest-high-level/src/main/java/org/elasticsearch/client/common/XContentSource.java:
```diff
@@ -77,9 +77,8 @@ public boolean isList() {
-    @SuppressWarnings("unchecked")
-    public <T> T getValue(String path) {
-        return (T) ObjectPath.eval(path, data);
+    public Object getValue(String path) {
```
revert

vamuzumd commented on libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ObjectPath.java:
```diff
@@ -37,9 +37,8 @@ private ObjectPath() {
-    @SuppressWarnings("unchecked")
-    public static <T> T eval(String path, Object object) {
-        return (T) evalContext(path, object);
+    public static Object eval(String path, Object object) {
```
revert

vamuzumd committed:
suppressed more unchecked warnings

vamuzumd committed:
Added wildcard for transform.builder

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

vamuzumd commented:
Hi @martijnvg , @pugnascotia , could you please review this PR?

pugnascotia commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/transform/chain/ExecutableChainTransform.java:
```diff
@@ -24,22 +24,27 @@
 
 public class ExecutableChainTransform extends ExecutableTransform<ChainTransform, ChainTransform.Result> {
 
+    @SuppressWarnings("rawtypes")
```
I think we should just put this annotation on the class.

pugnascotia commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java:
```diff
@@ -76,8 +77,8 @@ public boolean isMap() {
      */
-    public Map<String, Object> getAsMap() {
-        return (Map<String, Object>) data();
+    public Map<?, ?> getAsMap() {
```
I understand why you've made this change, but actually I think we want to keep this code the way it was. Can you revert this change please?

pugnascotia commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/support/xcontent/XContentSource.java:
```diff
@@ -90,8 +91,7 @@ public boolean isList() {
      */
-    public List<Object> getAsList() {
-        return (List<Object>) data();
+    public List<?> getAsList() { return ((List<?>) data());
```
Same here, can you revert this change please?

vamuzumd replied:
Reverted the changes in this file except one suppressed tag

pugnascotia reviewed (changes requested):
Thank you very much @vamuzumd, the PR needs a couple of tweaks but otherwise looks good.

vamuzumd committed:
feedback resolved

vamuzumd committed:
revereted XCOntentSource to master partially as per feedback

vamuzumd commented:
> Thank you very much @vamuzumd, the PR needs a couple of tweaks but otherwise looks good.

Thanks for the feedback @pugnascotia !

breskeby reviewed (approved):
lgtm apart from what @pugnascotia already mentioned

pugnascotia commented:
@elasticmachine test this please

pugnascotia commented:
@vamuzumd there are compilation failures. I suggest running `./gradlew precommit` to make sure the code all compiles and passes our style checks.

vamuzumd committed:
fixed few more warnings

pugnascotia commented:
Hi @vamuzumd, are you still interested in carrying this PR forwards?

vamuzumd commented:
> Hi @vamuzumd, are you still interested in carrying this PR forwards?

Sorry for the late, I will work towards this PR and send an update soon.

vamuzumd committed:
Fix warnings in Test code

vamuzumd committed:
Fix errors by capturing derived objects into base class, specifying input types for builder

elasticsearchmachine commented:
<!-- Automatically generated by [elasticsearchmachine], action: [checking labels] -->
<!-- This comment will be automatically removed when the labelling issues are resolved -->
Hello @vamuzumd, there are labelling issues with this PR that need to be fixed:

 * At least one version label is required
 * At least one change type label (starting with '>') is required

vamuzumd commented:
> @vamuzumd there are compilation failures. I suggest running `./gradlew precommit` to make sure the code all compiles and passes our style checks.

@pugnascotia, thanks! I just resolved all compilation and code check failures and pushed new changes, can you take a look?

pugnascotia commented:
@vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.

vamuzumd committed:
merge conflicts resolve

vamuzumd commented:
> @vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.

whoops, thanks for pointing out! Just resolved the conflicts.

vamuzumd commented:
> > @vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.
> 
> whoops, thanks for pointing out! Just resolved the conflicts. Working on build failures in  the pipeline.

vamuzumd committed:
fix warnings with latestConfig unchecked types

vamuzumd committed:
resolve SourceConfig unchecked

vamuzumd committed:
remove OpenFollwerIndexStepTests as deleted from source

vamuzumd commented:
> > > @vamuzumd would you mind fetching and merging `master`? There are some conflicts to resolve.
> > 
> > 
> > whoops, thanks for pointing out! Just resolved the conflicts. Working on build failures in  the pipeline.

Also resolve build failures. the PR looks ready for merge.

vamuzumd commented:
> Thank you very much @vamuzumd, the PR needs a couple of tweaks but otherwise looks good.

Thanks, @pugnascotia , I've resolved build issues and merge conflicts. Please lmk if the last iteration looks good.

vamuzumd committed:
more fixes from precommit

pugnascotia commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/config/RuleScope.java:
```diff
@@ -43,7 +44,7 @@
             for (Map.Entry<String, Object> entry : unparsedScope.entrySet()) {
                 try (XContentBuilder builder = XContentFactory.jsonBuilder()) {
-                    builder.map((Map<String, ?>) entry.getValue());
+                    builder.map((Map<String, Object>) entry.getValue());
```
`builder.map` expects a `Map<String, ?>`, so I don't think we need to change this line.

pugnascotia commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/watcher/actions/ActionWrapper.java:
```diff
@@ -174,7 +175,7 @@ public ActionWrapperResult execute(WatchExecutionContext ctx) {
                 int runs = 0;
                 if (object instanceof Collection) {
-                    Collection collection = Collection.class.cast(object);
+                    Collection<?> collection = Collection.class.cast(object);
```
I think in this case, we can just say:
```suggestion
                    Collection<?> collection = (Collection<?>) object;
```

pugnascotia commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/action/TransportXPackInfoActionTests.java:
```diff
@@ -43,6 +43,7 @@
 
 public class TransportXPackInfoActionTests extends ESTestCase {
 
+    @SuppressWarnings("unchecked")
```
The other option here, rather than suppressing the warning for the whole method, is to define a helper like so:

```java
    @SuppressWarnings("unchecked")
    private <T> ActionListener<T> anyActionListener() {
        return any(ActionListener.class);
    }
```

then replace `any(ActionListener.class)` with `anyActionListener()`. It's slightly annoying to have to do this, but it's an easy way to limit the scope of the suppressions.

vamuzumd replied:
That's a great suggestion! Unfortunately, that would still require adding suppression since the method does castings like  (ActionListener<XPackInfoFeatureResponse>)answer.getArguments()[2] on line 58. So, keeping this as is.

vamuzumd replied:
Plus it would require adding helpers for eq on line 57 and any(ActionRequest) which could get a bit gawky, but happy to change this if you feel otherwise.

pugnascotia commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/ClassificationTests.java:
```diff
@@ -89,6 +89,7 @@ protected NamedWriteableRegistry getNamedWriteableRegistry() {
         return new NamedWriteableRegistry(entries);
     }
 
+    @SuppressWarnings({"unchecked"})
```
I don't believe we need this here?

pugnascotia replied:
Turns out my IDE was lying - ignore me!

pugnascotia replied:
Ah, I see what the root problem is here. What we actually want to do is edit `ESTestCase.randomFrom` as follows. Then you won't need to add `@SuppressWarnings({"unchecked"})` to test methods that call `randomFrom()`.

```diff
diff --git a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
index a18fd0d6bcd..991550fe334 100644
--- a/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java
@@ -713,6 +713,8 @@ public abstract class ESTestCase extends LuceneTestCase {
     }
 
     /** Pick a random object from the given array. The array must not be empty. */
+    @SafeVarargs
+    @SuppressWarnings("varargs")
     public static <T> T randomFrom(T... array) {
         return randomFrom(random(), array);
     }
```

vamuzumd replied:
Grat suggestion! Significantly reduced number of files changed.

pugnascotia commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/RegressionTests.java:
```diff
@@ -82,6 +82,7 @@ public static Regression createRandom() {
         return createRandom(BoostedTreeParamsTests.createRandom());
     }
 
+    @SuppressWarnings({"unchecked"})
```
I don't believe we need this here?

pugnascotia commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/InferenceToXContentCompressorTests.java:
```diff
@@ -22,6 +22,7 @@
 
 import static org.hamcrest.Matchers.equalTo;
 
+@SuppressWarnings({"unchecked"})
```
I don't believe we need this here? That's what my IDE is telling me?

pugnascotia commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/NamedXContentObjectsTests.java:
```diff
@@ -147,6 +147,7 @@ public void chooseStrictOrLenient() {
         lenient = randomBoolean();
     }
 
+    @SuppressWarnings({"unchecked"})
```
This too, do we actually need this?

pugnascotia commented on x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAliasTests.java:
```diff
@@ -81,6 +81,7 @@
     private ArgumentCaptor<CreateIndexRequest> createRequestCaptor;
     private ArgumentCaptor<IndicesAliasesRequest> aliasesRequestCaptor;
 
+    @SuppressWarnings("unchecked")
```
We can use the `anyActionListener()` trick here again, as well as defining a `mockActionListener()` for the other compiler error.

vamuzumd replied:
Sure, we could do that, but like the other case, this requires defining 4 helpers plus address the casting on line 100, hence leaving this as is. But please lmk if you think otherwise.

pugnascotia commented:
@elasticmachine ok to test

pugnascotia commented:
@vamuzumd this is looking good, just a few things to fix up. I appreciate you tackling this work, some of the generics are...interesting 😱

vamuzumd committed:
resolve comments partial

vamuzumd committed:
eliminate unchecked suppressions for randomFrom by adding safevarargs

vamuzumd commented:
> @vamuzumd this is looking good, just a few things to fix up. I appreciate you tackling this work, some of the generics are...interesting 😱

Thanks @pugnascotia, I've addressed your feedback.

vamuzumd commented:
> @vamuzumd there are compilation failures. I suggest running `./gradlew precommit` to make sure the code all compiles and passes our style checks.



> > @vamuzumd this is looking good, just a few things to fix up. I appreciate you tackling this work, some of the generics are...interesting 😱
> 
> Thanks @pugnascotia, I've addressed your feedback.

@pugnascotia , the ci/bwc build fails for 2 tests in org.elasticsearch.backwards.MixedClusterClientYamlTestSuiteIT suite.
I don't see these failures associated with my changes as they belong to different project/package from XPack. Any suggestions?

vamuzumd committed:
Merge branch 'master' into vamuzumd/LowHangingFirst

vamuzumd commented:
Seems like the test BootStrapTests.testTriggeredWatchLoading failed for elasticsearch-ci/2, which seems flaky and unrelated to my change based on error. @pugnascotia, thoughts?

Partial stacktrace

java.lang.AssertionError: Count is 18 hits but 9 was expected.  Total shards: 1 Successful shards: 1 & 0 shard failures:
	at __randomizedtesting.SeedInfo.seed([76DF43B200299F27:4F3123083FC3A7D8]:0)
	at org.junit.Assert.fail(Assert.java:88)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount(ElasticsearchAssertions.java:259)
	at org.elasticsearch.xpack.watcher.test.integration.BootStrapTests.lambda$assertSingleExecutionAndCompleteWatchHistory$3(BootStrapTests.java:289)

vamuzumd commented:
@pugnascotia, could you please review/sign-off this PR? Just hoping to avoid another merge conflict since the PR has a large number of files changed.

pugnascotia commented:
Sorry @vamuzumd, I've been on vacation for the past week.

@elasticmachine update branch

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into vamuzumd/LowHangingFirst

pugnascotia reviewed (approved):
LGTM!

pugnascotia commented:
Thank you for your work @vamuzumd!

elasticsearchmachine commented:
<!-- Automatically generated by [elasticsearchmachine], action: [checking labels] -->
<!-- This comment will be automatically removed when the labelling issues are resolved -->
Hello @vamuzumd, there are labelling issues with this PR that need to be fixed:

 * At least one version label is required
 * At least one change type label (starting with '>') is required

pugnascotia commented:
@vamuzumd sorry for the noise, I should have added the right labels sooner.

pugnascotia commented:
Backported to `7.x` in 37a24d58721
